index.js

```diff
@@ -43,89 +60,234 @@ function cmdShim (src, to, opts) {
+ *
+ * @typedef {object} RuntimeInfo
+ * @property {string|null} [program] If `program` is `null`, the program may
+ * be a binary executable and can be called from shells by just its path.
+ * (e.g. `.\foo.exe` in CMD or PowerShell)
+ * @property {string} additionalArgs Additional arguments embedded in the shebang and passed to `program`.
+ * `''` if nothing, unlike `program`.
+ */
+
+/**
+ * Look into runtime (e.g. `node` & `sh` & `pwsh`) and its arguments
+ * of the target program (script or executable).
+ *
+ * @param {string} target Path to the executable or script.
+ * @return {Promise<RuntimeInfo>} Promise of infomation of runtime of `target`.
+ */
+function searchScriptRuntime (target) {
+  return fs
+    .readFile(target, 'utf8')
+    .then(data => {
+      // First, check if the bin is a #! of some sort.
+      const firstLine = data.trim().split(/\r*\n/)[0]
+      const shebang = firstLine.match(shebangExpr)
+      if (!shebang) {
+        // If not, infer script type from its extension.
+        // If the inferrence fails, it's something that'll be compiled, or some other
+        // sort of script, and just call it directly.
+        const targetExtension = path.extname(target).toLowerCase()
+        return Promise.resolve({
+          // undefined if extension is unknown but it's converted to null.
+          program: extensionToProgramMap.get(targetExtension) || null,
+          additionalArgs: ''
         })
-        .catch(() => writeShim_(src, to, Object.assign({}, opts, {args: defaultArgs})))
+      }
+      return Promise.resolve({
+        program: shebang[1],
+        additionalArgs: shebang[2]
+      })
     })
 }
 
-function writeShim_ (src, to, opts) {
-  opts = Object.assign({}, DEFAULT_OPTIONS, opts)
-  let shTarget = path.relative(path.dirname(to), src)
+/**
+ * Write shim to the file system while executing the pre- and post-processes
+ * defined in `WriteShimPre` and `WriteShimPost`.
+ *
+ * @param {string} src Path to the executable or script.
+ * @param {string} to Path to the (sh) shim(s) that is going to be created.
+ * @param {RuntimeInfo} srcRuntimeInfo Result of `await searchScriptRuntime(src)`.
+ * @param {ShimGenerator} generateShimScript Generator of shim script.
+ * @param {Options} opts Other options.
+ */
+function writeShim (src, to, srcRuntimeInfo, generateShimScript, opts) {
+  const defaultArgs = opts.preserveSymlinks ? '--preserve-symlinks' : ''
+  // `Array.prototype.flter` removes ''.
+  // ['--foo', '--bar'].join(' ') and [].join(' ') returns '--foo --bar' and '' respectively.
+  const args = [srcRuntimeInfo.additionalArgs, defaultArgs].filter(arg => arg).join(' ')
+  opts = Object.assign({}, opts, {
+    prog: srcRuntimeInfo.program,
+    args: args
+  })
+
+  return writeShimPre(to)
+    .then(() => fs.writeFile(to, generateShimScript(src, to, opts), 'utf8'))
+    .then(() => writeShimPost(to))
+}
+
+/**
+ * Callback functions to generate scripts for shims.
+ * @callback ShimGenerator
```

Comment:
These should be on the same line:

```suggestion
 * @callback ShimGenerator Callback functions to generate scripts for shims.
 *
```